### PR TITLE
VNNI Dialect and transform from linalg to VNNI 

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**

### CODE REVIEW

File: include/TPP/Dialect/VNNI/VNNIOps.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
+//===- VNNIOps.h - Check dialect ops -----------------------------*- C++
```
Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.h:1]: too long.

File: include/TPP/Dialect/VNNI/VNNIOps.td [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
```
Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.td:1]: same comment here.

File: include/TPP/Dialect/VNNI/VNNIOps.td [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CHECK_DIALECT_CHECK_OPS
```
Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.td:5]: please fix guard, it is not `CHECK` dialect.

File: include/TPP/Dialect/VNNI/VNNIOps.td [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  let cppNamespace = "::mlir::vnni";
+  let summary = [{
+    A dialect implementing vnni operations.
+  }];
+}
+
+def VNNI_MatmulOp : Op<VNNI_Dialect, "matmul"> {
+  let summary = [{Matmul operation with VNNI format of the first operand}];
+  let description = [{
+    Matmul Operation with VNNI format of the first operand..
```
Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.td:21]: double `.`.

Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.td:21]: The description read the same as the summary. Can you please improve? For example, can we add the access pattern of the operation (i.e., C[i][j] += A[i][k] * B[k][j]).

File: include/TPP/Dialect/VNNI/VNNIOps.td [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    ```mlir
+    vnni.matmul(%arg0, %arg1): memref<2x4x2xbf16>, memref<4x6xbf16>, memref<4x6xbf16>
+    ```
+  }];
+
+  let arguments = (ins BF16Tensor:$matrixA, BF16Tensor:$matrixB);
+  let results = (outs BF16Tensor:$matrixC);
+  let assemblyFormat = "`(` $matrixA `,` $matrixB `)`  attr-dict `:` type($matrixA) `,` type($matrixB) `,` type($matrixC) ";
```
Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.td:30]: I would make the IR a bit different: `vnni.matmul ins(%arg0: memref<2x4x2xbf16>, %arg1: memref<4x6xbf16>) -> memref<4x6xbf16>)`.

File: include/TPP/Dialect/VNNI/VNNIOps.td [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  }];
+}
+
+def VNNI_MatmulOp : Op<VNNI_Dialect, "matmul"> {
+  let summary = [{Matmul operation with VNNI format of the first operand}];
+  let description = [{
+    Matmul Operation with VNNI format of the first operand..
+
+    ```mlir
+    vnni.matmul(%arg0, %arg1): memref<2x4x2xbf16>, memref<4x6xbf16>, memref<4x6xbf16>
```
Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.td:24]: Why do we have two operands and not three?

File: include/TPP/Dialect/VNNI/VNNIOps.td [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+#ifndef VNNI_DIALECT_VNNI_OPS
+#define VNNI_DIALECT_VNNI_OPS
+
+include "mlir/IR/OpBase.td"
+
+def VNNI_Dialect : Dialect {
+  let name = "vnni";
+  let cppNamespace = "::mlir::vnni";
+  let summary = [{
+    A dialect implementing vnni operations.
```
Comment (rengolin) [inline include/TPP/Dialect/VNNI/VNNIOps.td:19]:
```suggestion
    A dialect implementing Intel Vector Neural Network Instructions operations.
    https://www.intel.com/content/www/us/en/artificial-intelligence/deep-learning-boost.html
```

File: include/TPP/Dialect/VNNI/VNNIOps.td [hunk 8]
```diff
@@ -15,19 +20,25 @@ def VNNI_Dialect : Dialect {
...
   let description = [{
-    Matmul Operation with VNNI format of the first operand..
+    Matmul Operation with VNNI format of the first operand.
+    For example, the operation performed is
+    C[i][j] += A[i/2][k][2]* B[k][j] with the first operand A's layout being
+    blocked by size 2.
 
     ```mlir
-    vnni.matmul(%arg0, %arg1): memref<2x4x2xbf16>, memref<4x6xbf16>, memref<4x6xbf16>
+    vnni.matmul(%arg0: memref<3x4x2xbf16>, %arg1: memref<4x6xbf16>)-> %arg2: memref<6x6xbf16>
```
Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.td:35]: Looks still a bit strange to me. What about:
``` vnni.matmul ins(%arg0: memref<3x4x2xbf16>, %arg1: memref<4x6xbf16>) out(%arg2: memref<6x6xbf16>)```

Comment (chelini) [reply include/TPP/Dialect/VNNI/VNNIOps.td:35]: Also can you please add a test along the line of: https://github.com/plaidml/tpp-sandbox/blob/main/test/TPP/tpp/tpp-ops.mlir

**Comment by rengolin** [comment]
LGTM, but I'll let @chelini approve to make a final pass. Thanks!

File: lib/TPP/ToBlockLayoutAndBack.cpp [hunk 9]
```diff
@@ -83,13 +84,32 @@ static Value handleLayoutNC_NCnc(Location loc, Value input, Value output,
...
+}
+
 // Helper function to pack from NC to NCnc.
 static Value toPackLayoutNC_NCnc(Location loc, Value input,
                                  ArrayRef<OpFoldResult> tiles,
                                  OpBuilder &builder, bool useAlloc = false) {
   return handleLayoutNC_NCnc(loc, input, nullptr, tiles, builder, useAlloc);
 }
 
+// Helper function to pack from NC to NCnc.
```
Comment (chelini) [inline lib/TPP/ToBlockLayoutAndBack.cpp:106]: Is this comment correct? If this is the case why we need another method?

File: include/TPP/Dialect/VNNI/VNNIOps.td [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+def VNNI_Dialect : Dialect {
+  let name = "vnni";
+  let cppNamespace = "::mlir::vnni";
+  let summary = [{
+    A dialect implementing Intel Vector Neural Network Instructions operations.
+    https://www.intel.com/content/www/us/en/artificial-intelligence/deep-learning-boost.html
+  }];
+}
+
+def VNNI_MatmulOp : Op<VNNI_Dialect, "matmul", [TypesMatchWith<"result type matches type of dest",
```
Comment (chelini) [inline include/TPP/Dialect/VNNI/VNNIOps.td:24]: can you format a bit better here? maybe `[TypesMatchWith<"result type matches type of dest"` this should go in a new line.

**Review by chelini — APPROVED** [review]

File: lib/TPP/ToBlockLayoutAndBack.cpp (hunk 9 above)
Comment (KavithaTipturMadhu) [reply lib/TPP/ToBlockLayoutAndBack.cpp:106]: Updated the comment, it uses a different inner_dim_pos as compared to NCnc.

**Pull request merged**